Prep for adding auto-rebalancing transactional producer.

danxmoran opened the pull request:
Part of #132.

I've got a WIP branch implementing the auto-rebalancing variant, but I'm still working out the concurrency details. The diff is big enough now that I thought it'd be worth pushing the pieces that refactor existing code.

I'll comment inline with explanations.

danxmoran commented on modules/core/src/main/scala/fs2/kafka/CommittableProducerRecords.scala:
```diff
@@ -34,26 +35,21 @@ import fs2.kafka.internal.syntax._
   * - `CommittableProducerRecords#one` to produce exactly one record within
   * the same transaction as the offset is committed.
   */
-sealed abstract class CommittableProducerRecords[F[_], G[+_], +K, +V] {
```
Trying to support `G[_]` here is a major pain when trying to batch records into transactions. The value-add seems to be allowing users to build instances of this class from semi-arbitrary collections of records. We can get the same value by keeping `G[_]` in the static builder methods, but then convert to `Chunk` right away to make downstream processing simpler.

vlovgr replied:
Sounds good! 👍

danxmoran commented on modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala:
```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
-                                groupId
-                              )
+                    consumerGroupId.flatMap { groupId =>
+                      records.flatTraverse { committable =>
```
This is one block where replacing `G[_]` with `Chunk` pays off, because it makes interleaving the various `traverse` + `flatten` combinators much easier.

danxmoran commented on modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala:
```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+        Resource
+          .make(settings.createProducer)(close(blocker))
+          .evalTap(init(blocker))
+          .map { producer => records =>
```
This is the other spot where removing `G[_]` pays off. Without the method-level type parameter we can pass the grouping logic around as a simple function. If we keep the `G[_]`, we have to introduce a new `trait` with a generic method and construct that here instead.

danxmoran commented on modules/core/src/main/scala/fs2/kafka/TransactionalProducerRecords.scala:
```diff
@@ -37,20 +37,20 @@ import fs2.kafka.internal.syntax._
   * commit exactly one offset, then emit a [[ProducerResult]] with the
   * results and specified passthrough value.
   */
-sealed abstract class TransactionalProducerRecords[F[_], G[+_], +K, +V, +P] {
```
As far as I can tell, the only point of this `G[_]` was to pass through the type of the wrapped `records` field.

danxmoran commented on modules/core/src/main/scala/fs2/kafka/package.scala:
```diff
@@ -257,7 +257,7 @@ package object kafka {
     context: ContextShift[F]
   ): Resource[F, TransactionalKafkaProducer[F, K, V]] =
-    TransactionalKafkaProducer.resource(settings)
+    new TransactionalKafkaProducer.Builder[F, K, V](settings).withConstantId
```
I expect we'll need to rethink the syntax helpers

vlovgr replied:
Yes, let's worry about that later on though. :)

danxmoran commented on modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala:
```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+      * only guaranteed by running separate producers with unique transactional
+      * IDs per topic/partition processed by the group.
+      */
+    def withConstantId: Resource[F, TransactionalKafkaProducer[F, K, V]] =
```
@vlovgr I _really_ want to make this take a `String` parameter to use as the constant ID and have this method inject the appropriate config, to entirely dodge the chance of a runtime error from passing in settings without a transactional ID, but then there isn't much point to having `withTransactionalId` on `ProducerSettings`. What do you think?

It makes even more sense in the dynamic case, where we need an arbitrary prefix for the rebalancing transactional IDs. Pulling the "transactional ID" out of settings and using it as the prefix works, but feels like it could be confusing to users. It also forces our code to deal with the error case when there's no ID present in config, vs. just throwing it at Kafka and catching any errors that bubble up.

vlovgr replied:
Agree, let's move transaction-specific configuration to these functions, rather than the Settings.

vlovgr commented on modules/core/src/main/scala/fs2/kafka/CommittableProducerRecords.scala:
```diff
@@ -67,8 +63,17 @@ object CommittableProducerRecords {
-  )(implicit G: Foldable[G]): CommittableProducerRecords[F, G, K, V] =
-    new CommittableProducerRecordsImpl(records, offset, G)
+  )(implicit G: Foldable[G]): CommittableProducerRecords[F, K, V] = {
+    val buf = new mutable.ArrayBuffer[ProducerRecord[K, V]]()
```
I think it's worth initializing this to `records.size`, as many `Foldable` have constant-time `size`.

vlovgr replied:
We can also specialize with `Chunk.empty` and `Chunk.singleton` on sizes 0 and 1, respectively. Especially since `one` uses `apply` (also possible to change `one` to `new CommittableProducerRecordsImpl(Chunk.singleton(record), offset)`).

danxmoran replied:
`size` on `Foldable` returns `Long`, but `ArrayBuffer`'s constructor expects `Int`. What's the best way to safely handle the overflow?

vlovgr replied:
I'd say it's unlikely you would want to produce more than `Int.MaxValue` records at a time, so `.toInt` sounds reasonable to me.

vlovgr commented on modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala:
```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+                ): F[ProducerResult[Chunk, K, V, P]] =
+                  grouper(records.records)
+                    .flatMap { recordGroups =>
+                      recordGroups.flatTraverse {
```
Should these run in parallel rather than sequentially?

danxmoran replied:
:+1: I think they should, we just need the extra type parameter.

vlovgr commented on modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala:
```diff
@@ -42,90 +42,124 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+        }
+      }
+
+    private def init(blocker: Blocker)(producer: KafkaByteProducer): F[Unit] =
```
You might want to make use of [`WithProducer`](https://github.com/ovotech/fs2-kafka/blob/master/modules/core/src/main/scala/fs2/kafka/internal/WithProducer.scala) in `TransactionalKafkaProducer`.

danxmoran replied:
I saw that, but I think the way it's set up will prevent manual setup / close management for `KafkaByteProducer`s that we'll need for the rebalancing case.

vlovgr replied:
> manual setup / close management

We'll need to have a `bracket` in there somewhere to ensure we `close` no matter what (e.g. on errors). But this can be done as a `Stream` with `Stream.bracket`.

danxmoran commented:
@vlovgr the last commit is a (pretty aggressive) refactor addressing the review comments. Moving the transaction parameters out of `ProducerSettings` broke all of the existing syntax helpers, so I deleted them for now. I'm 👍 to restoring them before this PR merges, I just wanted to reduce the noise while iterating on the core logic.

I also added a new test case for transaction timeouts. While working on it, I found that transaction timeouts end up raising `ProducerFencedException` instances. The doc for that error is:
```java
/**
 * This fatal exception indicates that another producer with the same <code>transactional.id</code> has been
 * started. It is only possible to have one producer instance with a <code>transactional.id</code> at any
 * given time, and the latest one to be started "fences" the previous instances so that they can no longer
 * make transactional requests. When you encounter this exception, you must close the producer instance.
 */
```
Emphasis _When you encounter this exception, you must close the producer instance_. Once I was poking around the javadocs I saw the same message listed on nearly every method as the only "safe" way to recover from an error mid-transaction.

Now that we've introduced an abstraction layer between `TransactionalKafkaProducer` and the underlying `KafkaByteProducer`, do you think it's worth trying to engineer an auto-restart of the underlying producer on errors? Or is it fine to bubble up errors and leave it to users to close & restart when desired?

danxmoran commented on modules/core/src/test/scala/fs2/kafka/BaseKafkaSpec.scala:
```diff
@@ -51,7 +56,10 @@ abstract class BaseKafkaSpec extends BaseAsyncSpec with EmbeddedKafka {
-        customBrokerProperties = Map("transaction.state.log.replication.factor" -> "1")
+        customBrokerProperties = Map(
+          "transaction.state.log.replication.factor" -> "1",
+          "transaction.abort.timed.out.transaction.cleanup.interval.ms" -> transactionTimeoutInterval.toMillis.toString
```
This is needed to force the broker to actually notice when open transactions time out.

vlovgr commented:
> do you think it's worth trying to engineer an auto-restart of the underlying producer on errors? Or is it fine to bubble up errors and leave it to users to close & restart when desired?

I think it would be useful to handle errors, similarly to how [`CommitRecovery`](https://github.com/ovotech/fs2-kafka/blob/master/modules/core/src/main/scala/fs2/kafka/CommitRecovery.scala) handles commit errors by default (although configurable). I think that can be implemented in a separate pull request though.

vlovgr commented on modules/core/src/main/scala/fs2/kafka/CommittableProducerRecords.scala:
```diff
@@ -67,8 +63,25 @@ object CommittableProducerRecords {
+        case Some(record) => Chunk.singleton(record)
+      }
+    } else {
+      val buf = new mutable.ArrayBuffer[ProducerRecord[K, V]](G.size(records).toInt)
```
`G.size(records).toInt` -> `numRecords.toInt`

vlovgr commented on modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala:
```diff
@@ -42,90 +45,142 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
+    ): Resource[F, TransactionalKafkaProducer[F, K, V]] =
+      Resource.liftF(settings.keySerializer).flatMap { keySerializer =>
+        Resource.liftF(settings.valueSerializer).flatMap { valueSerializer =>
+          settings.blocker.fold(Blockers.producer)(Resource.pure[F, Blocker]).flatMap { blocker =>
```
We should probably introduce a `Blocker.transactionalProducer`, which uses a cached thread-pool rather than a single executor thread pool, since we'll want multiple transactions to run in parallel. `Blockers.producer` will only allow one transaction to run at-a-time.

danxmoran commented:
@vlovgr do you want syntax bike-shedding to live in this PR, or would you rather wait until this and the rebalancing functionality are in? I have ideas, but I appreciate wanting to keep the PRs focused too 😄

If we're waiting on the syntax, what's the best way to deal with the mdoc errors?

vlovgr commented:
@danxmoran Yeah, let's do the syntax in a separate pull request. You can disable mdoc temporarily for a block by changing `scala mdoc:<modifier>`  to `scala` for the relevant blocks. Maybe leave a TODO so we don't forget to add it back.

Is this ready from your perspective? If so, I'll try to have a final lookthrough today or tomorrow, tidy up any last bits, and get it merged.

danxmoran commented:
@vlovgr I think this is ready, apart from the mdoc

vlovgr commented:
@danxmoran Sorry it's taken me a while longer to look through this properly. Happy to merge as-is. Thanks! 👍
